### ISSUE TITLE
"layer.data = layer.data" resets view in callbacks for 3D data

### ISSUE THREAD

**Issue by Karol-G** [napari/napari#3854]
## 🐛 Bug

Any callback will reset the view back to center if `layer.data = layer.data` is called within the callback for 3D data.

I had trouble of getting my labels updated in the view with the `@my_layer.mouse_move_callbacks.append` callback (for which I will open another issue) and found out that `layer.data = layer.data` updates the view, but also resets it to center.

## To Reproduce

1. Execute code
2. Click on "Activate bug" on right side
3. Hold Control+MoveMouseWheel
4. Then move the mouse or perform a click
5. The view will be reset

```python
from qtpy import QtWidgets
from napari_plugin_engine import napari_hook_implementation
import napari
import numpy as np

class MyWidget(QtWidgets.QWidget):
    def __init__(self, viewer: 'napari.viewer.Viewer') -> None:
        super().__init__()
        self.viewer = viewer
        self.btn = QtWidgets.QPushButton('Activate bug')
        self.btn.clicked.connect(self._on_click_init)

        self.setLayout(QtWidgets.QVBoxLayout())
        self.layout().addWidget(self.btn)

    def _on_click_init(self):
        my_layer = self.viewer.layers[0]

        @my_layer.mouse_move_callbacks.append
        def do_something1(layer, event):
            layer.data = layer.data

        @my_layer.mouse_drag_callbacks.append
        def do_something2(layer, event):
            layer.data = layer.data


@napari_hook_implementation
def napari_experimental_provide_dock_widget():
    return MyWidget


image = np.random.randint(low=0, high=255, size=(5, 10, 10))
viewer = napari.Viewer()
viewer.window.add_dock_widget(MyWidget(viewer))
viewer.add_image(image)
napari.run()
```

## Expected behavior

The view should not be reset.

## Environment

napari: 0.4.13.dev134+gfcde7328
Platform: Windows-10-10.0.19041-SP0
Python: 3.8.5 (default, Sep 3 2020, 21:29:08) [MSC v.1916 64 bit (AMD64)]
Qt: 5.9.7
PyQt5: 5.9.2
NumPy: 1.21.4
SciPy: 1.7.3
Dask: 2021.12.0
VisPy: 0.9.4

OpenGL:
- GL version: 4.6.0 - Build 27.20.100.8729
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1080, scale 1.0

Plugins:
- PartSeg Analysis project reader: 0.13.14
- PartSeg Image reader: 0.13.14
- PartSeg Image tiff writer: 0.13.14
- PartSeg Mask project reader: 0.13.14
- PartSeg Mask project writer: 0.13.14
- PartSeg Masked Tiff reader: 0.13.14
- PartSeg widgets: 0.13.14
- napari-medical-image-formats: 0.3.6
- napari-tools-menu: 0.1.7
- scikit-image: 0.4.13.dev134+gfcde7328
- svg: 0.1.5

Best
Karol

**Comment by jni** [comment]
This is probably the same bug as #3803. It's indeed a bad regression that we'll definitely fix before the next release. Thanks for the report @Karol-G!

**Comment by Karol-G** [comment]
Thanks, good too know.
And also just FYI: `layer.events.data()` also resets the view in 3D to center.
But that has probably the same cause as origin as the other ones.

**Comment by jni** [comment]
@Karol-G On my machine this is fixed in latest main by #3859! Please reopen if not fixed for you. Thanks for reporting!
